man/ifinfo.1
.\"  -*- nroff -*-
.\"
.\" Author: Kevin Rosenberg <kevin@ctsim.org>
.\"
.\" $Id: ifinfo.1,v 1.4 2002/05/03 10:28:17 kevin Exp $
.Dd April 19, 2000
.Dt IFINFO 1
.Os
.Sh NAME
.Nm ifinfo
.Nd Print IF (Image File) Information
.Sh SYNOPSIS
.Nm ifinfo iffile [OPTIONS]
.Sh DESCRIPTION 
.Pp
.Nm 
prints information about an IF file
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl Fl labels
Print history labels (default)
.It Fl Fl stats
Print image statistics (default) 
.It Fl Fl no-labels
Do not print history labels
.It Fl Fl no-stats
Do not print image statistics
.It Fl Fl debug
Set debug mode
.It Fl Fl verbose
Set verbose mode
.It Fl Fl version
Print version
.It Fl Fl help
Print terse help summary
.El
.Sh AUTHORS
Kevin Rosenberg, M.D. <kevin@ctsim.org>
.Sh HISTORY
CTSim was begun in 1983 using MS-DOS and an EGA display adapter. In
1999 it was ported to GNU/Linux and later ported to Microsoft Windows.
.Sh SEE ALSO
ctsim (1), if1 (1), if2 (1), ifexport (1), phm2pj (1), phm2if (1), pj2if (1), pjinfo (1), pjrec (1)


NEWS
Version 4.2 Released Mar 2003

* Direct Fourier Reconstructions now supported

Version 4.0 Released Jan 2003

* Port to wxWindows 2.4

* Port to Mac OS X


Version 3.5 Released May 2002

* New reconstruction technique: Interpolate divergent beams to
  parallel beams for faster reconstructions.

* Multithreading for running length processes in background. Also
  takes advantage of multiple-CPU's in SMP systems for speed improvements
  with parallel processing.

* Region of interest can be specified for reconstruction zooming.

* Import PNG, PPM, and PGM images

* DICOM file import/export

* Clipboard for cutting and pasting images

* Can read and reconstruct SIEMENS AR.STAR DICOM raw data files

* Scattergram of T-Theta sampling. Useful for understanding divergent beam
  scanning.

* Startup-tips to help new users learn features.

* Ability to place detector at any arbitray position

* Histogram plotting of projection files


Version 3.0 New Features

* Greatly improved dialog boxes

* Online and print manuals with context-sensitive help!

* 3-d rotating views of image files

* Creation of arbitrary filter images

* Cubic interpolation for reconstructions

* All features of command line tools are now in graphical ctsim program!

* Complex-valued image files now supported, lots of image math
 functions added including Fourier transformations

* Visual and statistical image comparision functions

* Plotting of row and column data of single and comparison images

* Histogram Plotting of Image files.

* Conversion of projections to polar images

* Improved animation graphics

* New Microsoft Windows compatible self-installer with explorer extensions 
to CTSim files



Version 2.5 of CTSim has been released!

New Features of CTSim version 2.5:

- Now compiles with Microsoft Visual C++.

- First Microsoft Windows graphical user interface

- Bug fixes and modest documentation and program improvements over version 2.0



New Features of CTSim version 2.0:

- Entire code-base cleaned up, re-written, and converted from C to C++.

- Graphical user interface now included! Animation of projection and
  reconstruction processes. Displays projection and image files. Uses
  platform-independent wxWindows library so can run on UNIX and 
  Microsoft Windows platforms.

- Integrated G2 library for graphical display from command line tools
  (UNIX only)

- Added frequency-based filtering. Can provide significant speed-up
  in reconstructions compared to convolution-based filtering.

- Added highly-optimized backprojection method (idiff3)

- New command-line tools added for image statistics and comparisons with
  both text and graphical outputs.


tools/ifinfo.cpp
/*****************************************************************************
** FILE IDENTIFICATION
**
**   Name:          ifinfo.cpp
**   Purpose:       Display information about an image file
**   Programmer:    Kevin Rosenberg
**   Date Started:  April 2000
**
**  This is part of the CTSim program
**  Copyright (C) 1983-2009 Kevin Rosenberg
**
**  This program is free software; you can redistribute it and/or modify
**  it under the terms of the GNU General Public License (version 2) as
**  published by the Free Software Foundation.
**
**  This program is distributed in the hope that it will be useful,
**  but WITHOUT ANY WARRANTY; without even the implied warranty of
**  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
**  GNU General Public License for more details.
**
**  You should have received a copy of the GNU General Public License
**  along with this program; if not, write to the Free Software
**  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
******************************************************************************/

/* FILE
 *   ifinfo.c             Display info on sdf files
 */

#include "ct.h"

enum { O_LABELS, O_STATS, O_NO_STATS, O_NO_LABELS, O_VERBOSE, O_HELP, O_VERSION, O_DEBUG };

static struct option my_options[] =
{
  {"labels", 0, 0, O_LABELS},
  {"no-labels", 0, 0, O_NO_LABELS},
  {"stats", 0, 0, O_STATS},
  {"no-stats", 0, 0, O_NO_STATS},
  {"debug", 0, 0, O_DEBUG},
  {"verbose", 0, 0, O_VERBOSE},
  {"help", 0, 0, O_HELP},
  {"version", 0, 0, O_VERSION},
  {0, 0, 0, 0}
};

static const char* g_szIdStr = "$Id$";


void
ifinfo_usage (const char *program)
{
  std::cout << "usage: " << fileBasename(program) << " image-filename [OPTIONS]\n";
  std::cout << "Imagefile information\n";
  std::cout << std::endl;
  std::cout << "     infile       Name of input IF file\n";
  std::cout << "     --display    Display image\n";
  std::cout << "     --labels     Print image labels (default)\n";
  std::cout << "     --no-labels  Do not print image labels\n";
  std::cout << "     --stats      Print image statistics (default)\n";
  std::cout << "     --no-stats   Do not print image statistics\n";
  std::cout << "     --debug      Debug mode\n";
  std::cout << "     --verbose    Verbose mode\n";
  std::cout << "     --version    Print version\n";
  std::cout << "     --help       Print this help message\n";
}

int
ifinfo_main (int argc, char *const argv[])
{
  ImageFile *im = NULL;
  std::string in_file;
  int opt_verbose = 0;
  int opt_stats = 1;
  int opt_labels = 1;
  int opt_debug = 0;

  while (1)
    {
      int c = getopt_long (argc, argv, "", my_options, NULL);

      if (c == -1)
        break;

      switch (c)
        {
        case O_LABELS:
          opt_labels = 1;
          break;
        case O_STATS:
          opt_stats = 1;
          break;
        case O_NO_LABELS:
          opt_labels = 0;
          break;
        case O_NO_STATS:
          opt_stats = 0;
          break;
        case O_VERBOSE:
          opt_verbose = 1;
          break;
        case O_DEBUG:
          opt_debug = 0;
          break;
        case O_VERSION:
#ifdef VERSION
          std::cout << "Version " << VERSION << std::endl << g_szIdStr << std::endl;
#else
          std::cout << "Unknown version number\n";
#endif
          return (0);
        case O_HELP:
        case '?':
          ifinfo_usage(argv[0]);
          return (0);
        default:
          ifinfo_usage(argv[0]);
          return (1);
        }
    }

  if (optind + 1 != argc) {
    ifinfo_usage (argv[0]);
    return (1);
  }

  in_file = argv[optind];

  im = new ImageFile ();
  if (! im->fileRead (in_file)) {
    sys_error (ERR_WARNING, "Unable to read file %s", in_file.c_str());
    return (1);
  }

  if (opt_labels)
    im->printLabels (std::cout);

  if (opt_stats) {
    std::cout << "Size: (" << im->nx() << "," << im->ny() << ")\n";
    std::cout << "Data type: ";
    if (im->dataType() == Array2dFile::DATA_TYPE_COMPLEX)
      std::cout << "Complex\n";
    else
      std::cout << "Real\n";

    im->printStatistics (std::cout);
  }

  return (0);
}

#ifndef NO_MAIN
int
main (int argc, char *const argv[])
{
  int retval = 1;

  try {
    retval = ifinfo_main(argc, argv);
  } catch (exception e) {
          std::cerr << "Exception: " << e.what() << std::endl;
  } catch (...) {
          std::cerr << "Unknown exception\n";
  }

  return (retval);
}
#endif
